server/src/com/thoughtworks/go/server/service/EntityHashingService.java

```diff
@@ -109,6 +110,14 @@ public void onEntityConfigChange(SCM scm) {
         }
     }
 
+    class TemplateConfigChangedListner extends EntityConfigChangedListener<PipelineTemplateConfig> {
+        @Override
+        public void onEntityConfigChange(PipelineTemplateConfig pipelineTemplateConfig) {
+            String cacheKey =  pipelineTemplateConfig.getClass().getName() + pipelineTemplateConfig.name().toLower();
+            goCache.remove(ETAG_CACHE_KEY, cacheKey.toLowerCase());
```

Comment:
This could be problematic — if there is an environment, pipeline, scm and template which share the same name, you'd see cache key collision.

Probably change the cache key to `ETAG_CACHE_KEY + configObject.getClass() + configObject.name().toLower()`

Could you fix them in a separate PR, and fix this instance here?